co-noir/ultrahonk/src/parse/types.rs

```diff
@@ -0,0 +1,1259 @@
+use super::builder::{GenericUltraCircuitBuilder, UltraCircuitBuilder, UltraCircuitVariable};
+use super::plookup::{BasicTableId, MultiTableId};
+use crate::decider::polynomial::Polynomial;
+use crate::types::ProvingKey;
+use crate::Utils;
+use ark_ec::pairing::Pairing;
+use ark_ff::{One, PrimeField, Zero};
+use itertools::izip;
+use num_bigint::BigUint;
+use std::array;
+use std::cmp::Ordering;
+use std::collections::{HashMap, HashSet};
+use std::hash::{Hash, Hasher};
+use std::ops::{Index, IndexMut};
+
+#[derive(Default, PartialEq, Eq)]
+pub(crate) struct PolyTriple<F: PrimeField> {
+    pub(crate) a: u32,
+    pub(crate) b: u32,
+    pub(crate) c: u32,
+    pub(crate) q_m: F,
+    pub(crate) q_l: F,
+    pub(crate) q_r: F,
+    pub(crate) q_o: F,
+    pub(crate) q_c: F,
+}
+
+#[derive(Default, PartialEq, Eq)]
+pub(crate) struct AddTriple<F: PrimeField> {
+    pub(crate) a: u32,
+    pub(crate) b: u32,
+    pub(crate) c: u32,
+    pub(crate) a_scaling: F,
+    pub(crate) b_scaling: F,
+    pub(crate) c_scaling: F,
+    pub(crate) const_scaling: F,
+}
+
+#[derive(Default, PartialEq, Eq)]
+pub(crate) struct AddQuad<F: PrimeField> {
+    pub(crate) a: u32,
+    pub(crate) b: u32,
+    pub(crate) c: u32,
+    pub(crate) d: u32,
+    pub(crate) a_scaling: F,
+    pub(crate) b_scaling: F,
+    pub(crate) c_scaling: F,
+    pub(crate) d_scaling: F,
+    pub(crate) const_scaling: F,
+}
+
+#[derive(Default, PartialEq, Eq)]
+pub(crate) struct MulQuad<F: PrimeField> {
+    pub(crate) a: u32,
+    pub(crate) b: u32,
+    pub(crate) c: u32,
+    pub(crate) d: u32,
+    pub(crate) mul_scaling: F,
+    pub(crate) a_scaling: F,
+    pub(crate) b_scaling: F,
+    pub(crate) c_scaling: F,
+    pub(crate) d_scaling: F,
+    pub(crate) const_scaling: F,
+}
+
+pub(crate) struct MemOp<F: PrimeField> {
+    pub(crate) access_type: u8,
+    pub(crate) index: PolyTriple<F>,
+    pub(crate) value: PolyTriple<F>,
+}
+
+#[derive(PartialEq, Eq)]
+#[allow(clippy::upper_case_acronyms)]
+pub(crate) enum BlockType {
+    ROM = 0,
+    RAM = 1,
+    CallData = 2,
+    ReturnData = 3,
+}
+
+impl Default for BlockType {
+    fn default() -> Self {
+        Self::ROM
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct BlockConstraint<F: PrimeField> {
+    pub(crate) init: Vec<PolyTriple<F>>,
+    pub(crate) trace: Vec<MemOp<F>>,
+    pub(crate) type_: BlockType,
+    pub(crate) calldata: u32,
+}
+
+#[derive(Default)]
+pub(crate) struct AcirFormatOriginalOpcodeIndices {
+    // pub(crate)logic_constraints: Vec<usize>,
```

Comment:
I guess these are all of the black-box calls that are supported by barettenberg. Seems like a good idea to keep this here as a way to see the current support.

Reply:
Yes, I copied the struct definition and commented out everything I don't have yet.